containers/src/Data/IntMap/Internal.hs

```diff
@@ -769,6 +769,8 @@ disjoint t1@(Bin p1 m1 l1 r1) t2@(Bin p2 m2 l2 r2)
 -- 'compose' that forced the values of the output 'IntMap'. This version does
 -- not force these values.
 --
+-- __Note:__ This is not asymptotically optimal. See note at 'Data.Map.compose', keeping in mind \( \log(m) \) is \( O(\min(m,W)) \), but not the other way around.
```

Comment:
Actually, I'm not sure what would even be asymptotically optimal here. The current impl is `O(n * min(m,W)`, but one could write

```haskell
compose :: IntMap c -> IntMap Int -> IntMap c
compose bc !ab
  | null bc = empty
  | otherwise = mapMaybe (f bc !?) ab
  where
    f = Map.fromAscList . IntMap.toAscList
```

It seems to me this would be `O(n * log m + m)`, and the extra `m` makes some sense given we can't directly use element comparisons on lookup on `IntMap`, so either we convert to `Map` or rely on the word size bound.

No clue how the constants here play out, would need to benchmark.
I also wonder if there's a faster way to implement `f` as I couldn't find any Map/IntMap conversion functions in their respective modules.

Reply:
The question of optimality for `IntMap` doesn't seem to be well-posed, so it's probably not worth commenting on it. The cost of `compose`  is the cost of `n` lookups, that seems a satisfactory enough answer so that formalizing the question of optimality is unnecessary, if possible at all..

It's easy to talk about the optimality of `Map` because it matches the information-theoretic lower bound, which is independent of the representation, that makes it easy to formalize. For `IntMap`, a meaningful lower bound would have to take into account the particular data type involved, that seems difficult to formalize.